### PR TITLE
BuildSystem delegate API

### PR DESCRIPTION
**Pull request by krzyzanowskim**
BuildSystem progress notification API via delegate

### Motivation:

The `BuildOperation` acts as a black box, that makes it impossible to get meaningful notifications about the build events.

forum thread: https://forums.swift.org/t/how-buildsystem-api-is-accessible-for-the-clients/43787

### Modifications:

Introduce `BuildSystemDelegate` that can be used to track the progress of the build operation by the clients.

### Result:

- add optional `delegate` to `SPMBuildCore.BuildSystem`
- implement delegate calls for `BuildOperation` and `XcodeBuildSystem`

### CODE REVIEW

**Comment by abertelrud** [comment]
This looks like a good addition to me!  One question:  what are the guarantees about when the delegate methods will be called, e.g. are they guaranteed to be called all on the same queue so that there won't be any overlap between them?  I'm not that familiar with the concurrency behavior of the callbacks from llbuild; perhaps that's already part of the contract, but if not, would it make sense to have a serial queue that they're all invoked on so that ever client doesn't have to implement its own concurrency control?

**Comment by krzyzanowskim** [comment]
> what are the guarantees about when the delegate methods will be called, e.g. are they guaranteed to be called all on the same queue so that there won't be any overlap between them?

the delegate `queue` is already there (frankly, I don't use it, but will by the time I'll finish PR).

**Comment by abertelrud** [comment]
> > what are the guarantees about when the delegate methods will be called, e.g. are they guaranteed to be called all on the same queue so that there won't be any overlap between them?
> 
> the delegate `queue` is already there (frankly, I don't use it, but will by the time I'll finish PR).

Great, thanks!  Didn't see that there was a queue already; I was only looking at the new callsites.  Thanks!

**Comment by tomerd** [comment]
@krzyzanowskim is this one ready for review?

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Comment by krzyzanowskim** [comment]
@tomerd yes, please do.

**Review by abertelrud — APPROVED** [review]
This looks good to me; thank you for the additions!  I had a couple of questions but they are minor.

**Comment by abertelrud** [comment]
It looks as if the failures are because the CMake scripts need to be updated (for the new files).

**Comment by tomerd** [comment]
@krzyzanowskim lets get CMake fixed up and add the relevant comments so we can merge this

File: Sources/Build/BuildOperation.swift [hunk 1]
```diff
@@ -29,7 +32,7 @@ public final class BuildOperation: PackageStructureDelegate, SPMBuildCore.BuildS
     let packageGraphLoader: () throws -> PackageGraph
 
     /// The build delegate reference.
```
[the remark this answers is not in the document]
Comment (krzyzanowskim) [reply Sources/Build/BuildOperation.swift]: naming is hard indeed, especially that "build system" is ambiguous in the context. There's llbuild build system, and there's SPM build system. Would that work: 7e4f0bd25482efaeec439e356c39723593525fbf ?

**Review by neonichu — APPROVED** [review]

**Comment by tomerd** [comment]
@swift-ci please test

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Comment by tomerd** [comment]
@abertelrud please confirm changes look good based on you previous feedback so we can merge this

File: Sources/Build/BuildOperation.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (abertelrud) [reply Sources/Build/BuildOperation.swift]: Yes, that's clearer.  Thanks!

**Comment by abertelrud** [comment]
> @abertelrud please confirm changes look good based on you previous feedback so we can merge this

Thanks @tomerd they look good to me.  As far as I am concerned this can be merged.

**Comment by abertelrud** [comment]
@swift-ci please smoke test

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Comment by abertelrud** [comment]
Looks like SourceKit-LSP is affected:
```
10:20:48 
/Users/buildnode/jenkins/workspace/swift-package-manager-PR-macos-smoke-test/branch-main/sourcekit-lsp/Sources/SKSwiftPMWorkspace/SwiftPMWorkspace.swift:47:29: error: 'BuildSystemDelegate' is ambiguous for type lookup in this context
10:20:48   public weak var delegate: BuildSystemDelegate? = nil
10:20:48                             ^~~~~~~~~~~~~~~~~~~
10:20:48 /Users/buildnode/jenkins/workspace/swift-package-manager-PR-macos-smoke-test/branch-main/swiftpm/Sources/SPMBuildCore/BuildSystemDelegate.swift:14:17: note: found this candidate
10:20:48 public protocol BuildSystemDelegate: AnyObject {
10:20:48                 ^
10:20:48 /Users/buildnode/jenkins/workspace/swift-package-manager-PR-macos-smoke-test/branch-main/sourcekit-lsp/Sources/SKCore/BuildSystemDelegate.swift:17:17: note: found this candidate
10:20:48 public protocol BuildSystemDelegate: AnyObject {
10:20:48                 ^
```

**Comment by abertelrud** [comment]
Not sure if there is an interim change that can be put into SourceKit-LSP to avoid the ambiguity, or whether this PR needs to change.  @benlangmuir what would you suggest?

**Comment by benlangmuir** [comment]
https://github.com/apple/sourcekit-lsp/pull/370

@swift-ci please smoke test

**Comment by benlangmuir** [comment]
> Not sure if there is an interim change that can be put into SourceKit-LSP to avoid the ambiguity, or whether this PR needs to change. @benlangmuir what would you suggest?

I changed sourcekit-lsp (apple/sourcekit-lsp#370) to use a fully qualified reference, which seemed to be sufficient.

**Comment by krzyzanowskim** [comment]
4wvy0h

**Pull request merged**